slogr/slogr.go

```diff
@@ -0,0 +1,126 @@
+/*
+Copyright 2023 The logr Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package slogr provides bridges between github.com/go-logr/logr and Go's
+// slog package.
+package slogr
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	"golang.org/x/exp/slog"
+)
+
+// New returns a logr.Logger which logs through an arbitrary slog.Handler.
+func New(handler slog.Handler) logr.Logger {
+	return logr.New(newSink(slog.New(handler)))
+}
+
+// Underlier exposes access to the underlying logging function. Since
+// callers only have a logr.Logger, they have to know which
+// implementation is in use, so this interface is less of an
+// abstraction and more of a way to test type conversion.
+type Underlier interface {
+	//FIXME: should this return the Logger or Handler?
+	GetUnderlying() slog.Handler
+}
+
+func newSink(slogger *slog.Logger) logr.LogSink {
+	sink := &slogSink{
+		slogger: slogger,
+	}
+	/* // FIXME:
+	// For skipping slogSink.Info and slogSink.Error.
+	sink.Formatter.AddCallDepth(1)
+	*/
+	return sink
+}
+
+// slogSink inherits some of its LogSink implementation from Formatter
+// and just needs to add some glue code.
+type slogSink struct {
+	slogger *slog.Logger
+	name    string
+}
+
+// Init configures this Formatter from runtime info, such as the call depth
+// imposed by logr itself.
+// Note that this receiver is a pointer, so depth can be saved.
+func (sink *slogSink) Init(info logr.RuntimeInfo) {
+	//f.depth += info.CallDepth
+}
+
+// Enabled checks whether an info message at the given level should be logged.
+func (sink slogSink) Enabled(level int) bool {
+	return sink.slogger.Enabled(context.TODO(), slog.Level(-level))
+}
+
+func (sink slogSink) WithName(name string) logr.LogSink {
+	if len(sink.name) > 0 {
+		sink.name += "/"
+	}
+	sink.name += name
+	return &sink
+}
+
+func (sink slogSink) WithValues(kvList ...interface{}) logr.LogSink {
+	newSlogger := sink.slogger.With(kvList...)
+	return &slogSink{
+		slogger: newSlogger,
+		name:    sink.name,
+	}
+}
+
+/*
+// FIXME: this is impossible with slog, as far as I can tell
+// https://cs.opensource.google/go/x/exp/+/d63ba01a:slog/logger.go;l=240
```

Comment:
It is impossible with `slog.Logger`, but it can be done with `slog.Handler`. That's the key difference between this implementation of a `logr.LogSink` and the one in https://github.com/go-logr/logr/pull/196: mine wraps an `slog.Handler`.